exchanges/gateio/gateio.go

```diff
@@ -424,169 +1013,2938 @@ func (g *Gateio) SendAuthenticatedHTTPRequest(ctx context.Context, ep exchange.U
+	if ccy.IsEmpty() {
+		return nil, currency.ErrCurrencyCodeEmpty
+	}
+	var response CurrencyAndAmount
+	params := url.Values{}
+	params.Set("currency", ccy.String())
+	return &response, g.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, http.MethodGet, crossMarginTransferable, params, nil, &response)
+}
+
+// GetMaxBorrowableAmountForSpecificCrossMarginCurrency returns the max borrowable amount for a specific cross margin currency
+func (g *Gateio) GetMaxBorrowableAmountForSpecificCrossMarginCurrency(ctx context.Context, ccy currency.Code) (*CurrencyAndAmount, error) {
+	if ccy.IsEmpty() {
+		return nil, currency.ErrCurrencyCodeEmpty
+	}
+	var response CurrencyAndAmount
+	params := url.Values{}
+	params.Set("currency", ccy.String())
+	return &response, g.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, http.MethodGet, crossMarginBorrowable, params, nil, &response)
+}
+
+// GetCrossMarginBorrowHistory retrives cross margin borrow history sorted by creation time in descending order by default.
+// Set reverse=false to return ascending results.
+func (g *Gateio) GetCrossMarginBorrowHistory(ctx context.Context, status uint, currency currency.Code, limit, offset uint64, reverse bool) ([]CrossMarginLoanResponse, error) {
+	params := url.Values{}
+	if status < 1 || status > 3 {
+		return nil, fmt.Errorf("%s %v, only allowed status values are 1:failed, 2:borrowed, and 3:repayment", g.Name, errInvalidOrderStatus)
+	}
+	params.Set("status", strconv.Itoa(int(status)))
+	if !currency.IsEmpty() {
+		params.Set("currency", currency.String())
+	}
+	if limit > 0 {
+		params.Set("limit", strconv.FormatUint(limit, 10))
+	}
+	if offset > 0 {
+		params.Set("offset", strconv.FormatUint(offset, 10))
+	}
+	if reverse {
+		params.Set("reverse", strconv.FormatBool(reverse))
+	}
+	var response []CrossMarginLoanResponse
+	return response, g.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, http.MethodGet, crossMarginLoans, params, nil, &response)
+}
+
+// GetSingleBorrowLoanDetail retrieve single borrow loan detail
+func (g *Gateio) GetSingleBorrowLoanDetail(ctx context.Context, loanID string) (*CrossMarginLoanResponse, error) {
+	if loanID == "" {
+		return nil, errInvalidLoanID
+	}
+	var response CrossMarginLoanResponse
+	return &response, g.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, http.MethodGet, crossMarginLoans+"/"+loanID, nil, nil, &response)
+}
+
+// *********************************Futures***************************************
+
+// GetAllFutureContracts  retrives list all futures contracts
+func (g *Gateio) GetAllFutureContracts(ctx context.Context, settle string) ([]FuturesContract, error) {
+	var contracts []FuturesContract
+	settle = strings.ToLower(settle)
+	if settle != settleBTC && settle != settleUSD && settle != settleUSDT {
+		return nil, errMissingSettleCurrency
+	}
+	return contracts, g.SendHTTPRequest(ctx, exchange.RestSpot,
+		fmt.Sprintf(futuresSettleContracts, settle), &contracts)
```

Comment:
if possible please use concatenation instead of sprintf across this implementation.